src/ORM/EagerLoadedDataList.php

```diff
@@ -0,0 +1,893 @@
+<?php
+
+namespace SilverStripe\ORM;
+
+use SilverStripe\View\ViewableData;
+use SilverStripe\ORM\FieldType\DBField;
+use BadMethodCallException;
+use InvalidArgumentException;
+use LogicException;
+use SilverStripe\Dev\Debug;
+use SilverStripe\ORM\Connect\DatabaseException;
+use Traversable;
+
+/**
+ * Represents an "eager loaded" DataList - i.e. the data has already been fetched from the database
+ * for these records and likely for some of their relations.
+ *
+ * This list is designed to be plug-and-play with the various DataList implementations, with the exception
+ * that because it doesn't make a database query to get its data, some methods are intentionally not implemented.
+ *
+ * Note that when this list represents a relation, adding items to or removing items from this list _will not_
+ * affect the underlying relation in the database. This list is read-only.
+ */
+class EagerLoadedDataList extends ViewableData implements Relation, SS_List, Filterable, Sortable, Limitable
+{
+    /**
+     * List responsible for instantiating the actual DataObject objects from eager-loaded data
+     */
+    private DataList $dataList;
+
+    /**
+     * Underlying DataObject class for this list
+     */
+    private string $dataClass;
+
+    /**
+     * The ID(s) of the record that owns this list if the list represents a relation
+     * Used for aggregations
+     */
+    private int|array|null $foreignID;
+
+    /**
+     * ID-indexed array that holds the data from SQL queries for the list
+     * @var array<int,array>
+     */
+    private array $rows = [];
+
+    /**
+     * Nested eager-loaded data which applies to relations on records contained in this list
+     * @var array<int,self|DataObject>
```

Comment:
Note that this `array<{keyType},{valueType}>` syntax is very useful for IDEs to declare both the type for the _key_ as well as the value. It isn't technically in [the PHPDoc spec](https://docs.phpdoc.org/guide/references/phpdoc/types.html#arrays) but seems to be widely supported and is even used by tools such as phpstan.
If there's hesitance to use this syntax I can change it to comply with the spec.